lib/cmap/connection.js

```diff
@@ -80,19 +113,47 @@ class Connection extends EventEmitter {
     });
   }
 
+  // Wire protocol methods
   command(ns, cmd, options, callback) {
-    // NOTE: The wire protocol methods will eventually be migrated to this class, but for now
-    //       we need to pretend we _are_ a server.
-    const server = {
-      description: this.description,
-      s: {
-        bson: this.bson,
-        pool: { write: write.bind(this) }
-      }
-    };
+    wp.command(makeServerTrampoline(this), ns, cmd, options, callback);
+  }
 
-    wp.command(server, ns, cmd, options, callback);
+  query(ns, cmd, cursorState, options, callback) {
```

Comment:
From a design perspective I understand that the delegation is most likely here for convenience but I find it a bit odd that a generic connection object would know anything about creating specific wire protocol commands, especially when there's a circular dependency between the wire protocol helpers and the connection. My expectation would be that the creation of the wire protocol messages would be handled before passing them to the connection, and the connection itself would ust know how to send messages that adhered to a common interface. But I'm happy to have that discussion in another refactoring task.

Reply:
I agree, but we have to take a very long road to get there. The reason there is a circular dependency between the `Connection` and the wire protocol messages is only because we have to share code between the legacy pool and this one for now. Beyond that, you are correct that it just becomes a matter of personal preference for how you want to deal with a connection in operation execution:
```
class FindOperation {
   execute(conn, callback) {
     if (conn.maxWireVersion < 4) {
        conn.write(new LegacyFindWireProtocolMessage(...), callback);
        return;
     }

     conn.write(new CommandWireProtocolMessage(new FindCommand(), callback);
   }
}
```
vs
```
class FindOperation {
   execute(conn, callback) {
      conn.command(new FindCommand(), callback);
   }
}
```
where `conn.command` is just a helper that looks like:
```
class Connection {
    command(cmd, callback) {
      if (this.maxWireVersion < 4) {
        this.write(new LegacyFindWireProtocolMessage(...), callback);
        return;
      }

      this.write(new FindCommand(..), callback);
   }
}
```

Until we cut the old crufty topology layer, it costs a lot less to continue to pretend a `Connection` is a `Server`, and not touch any of the internals of the "wire protocol layer"

Reply:
Sounds good to me.